lib/consumer/message_tracker/server.ex

```diff
@@ -1,222 +1,159 @@
 
+  @impl true
   def init(init_args) do
-    acknowledger = Keyword.fetch!(init_args, :acknowledger)
-    {:ok, {:new, :queue.new(), %{}, acknowledger}}
-  end
-
-  def handle_call(:register_data_updater, from, {:new, queue, ids, acknowledger}) do
-    monitor(from)
-    {:reply, :ok, {:accepting, queue, ids, acknowledger}}
-  end
-
-  def handle_call(:register_data_updater, from, {_state, queue, ids, acknowledger}) do
-    Logger.debug("Blocked data updater registration. Queue is #{inspect(queue)}.")
-
-    {:noreply, {{:waiting_cleanup, from}, queue, ids, acknowledger}}
+    sharding_key = Keyword.fetch!(init_args, :sharding_key)
+    state = %State{queue: :queue.new(), sharding_key: sharding_key, data_updater_pid: nil}
+    {:ok, state, @message_tracker_deactivation_interval_ms}
   end
 
-  def handle_call(
-        {:can_process_message, message_id},
-        from,
-        {:accepting, queue, ids, acknowledger} = s
-      ) do
-    case :queue.peek(queue) do
-      {:value, ^message_id} ->
-        case Map.get(ids, message_id) do
-          nil ->
-            {:noreply, {{:waiting_delivery, from}, queue, ids, acknowledger}}
-
-          {:requeued, _delivery_tag} ->
-            {:noreply, {{:waiting_delivery, from}, queue, ids, acknowledger}}
-
-          _ ->
-            {:reply, true, s}
-        end
-
-      {:value, _} ->
-        {:reply, false, s}
-
-      :empty ->
-        Logger.debug("#{inspect(message_id)} has not been tracked yet. Waiting.")
-        {:noreply, {{:waiting_delivery, message_id, from}, queue, ids, acknowledger}}
+  @impl true
+  def handle_cast({:handle_message, %Message{} = message, channel}, state) do
+    # :queue.len/1 runs in O(n), :queue.is_empty in O(1)
+    if :queue.is_empty(state.queue) do
+      new_state = put_message_in_queue(message, channel, state)
+      {:noreply, new_state, {:continue, :process_message}}
+    else
+      new_state = put_message_in_queue(message, channel, state)
+      {:noreply, new_state, @message_tracker_deactivation_interval_ms}
     end
   end
 
-  def handle_call({:ack_delivery, message_id}, _from, {:accepting, queue, ids, acknowledger}) do
-    {{:value, ^message_id}, new_queue} = :queue.out(queue)
-    {delivery_tag, new_ids} = Map.pop(ids, message_id)
+  @impl true
+  def handle_call({:ack_delivery, %Message{} = message}, _from, state) do
+    # Invariant: we're always processing the first message in the queue
+    # Let us make sure
+    case :queue.peek(state.queue) do
+      {:value, %QueueEntry{message: ^message, channel: channel}} ->
+        @adapter.ack(channel, message.meta.delivery_tag)
+        new_state = drop_head_from_queue(state)
+        # let's move on to the next message
+        {:reply, :ok, new_state, {:continue, :process_message}}
 
-    :ok = ack(acknowledger, delivery_tag)
-
-    {:reply, :ok, {:accepting, new_queue, new_ids, acknowledger}}
+      _ ->
+        # discard the message
+        {:reply, :ok, state, {:continue, :process_message}}
```

Comment:
We want to continue to `:process_message` in order to move on with the next message in the queue. Maybe it would be clearer if the state was called `:next_message`?
The fact that the message received via `ack_received` is not acked is enough to discard it, as we do nothing more with it